cpp/arduino/Wire.h

```diff
@@ -1,138 +1,228 @@
-  void beginTransmission(int address) {
-    // TODO: implement
-  }
+  // Begin a transmission to the I2C slave device with the given address.
+  // Subsequently, queue bytes for transmission with the write() function and
+  // transmit them by calling endTransmission().
+  // For the mock we update our output to the proper destination.
   void beginTransmission(uint8_t address) {
-    beginTransmission((int)address);
+    assert(_didBegin);
+    assert(address > 0 && address < SLAVE_COUNT);
+    assert(out == nullptr);
+    out = &slaves[address];
+    out->mosiSize = 0;
   }
+  void beginTransmission(int address) { beginTransmission((uint8_t)address); }
 
   // https://www.arduino.cc/en/Reference/WireEndTransmission
-  // Ends a transmission to a slave device that was begun by beginTransmission() and transmits the bytes that were
-  // queued by write().
-  uint8_t endTransmission(uint8_t sendStop) {
-    // TODO: implement
+  // Ends a transmission to a slave device that was begun by beginTransmission()
+  // and transmits the bytes that were queued by write().
+  // In the mock we just leave the bytes there in the buffer
+  // to be read by the testing API and we ignore the sendStop.
+  uint8_t endTransmission(bool sendStop) {
+    assert(_didBegin);
+    assert(out);
+    out = nullptr;
     return 0; // success
   }
-  uint8_t endTransmission(void) {
-    return endTransmission((uint8_t)true);
-  }
+  uint8_t endTransmission(void) { return endTransmission(true); }
 
   // https://www.arduino.cc/en/Reference/WireRequestFrom
-  // Used by the master to request bytes from a slave device. The bytes may then be retrieved with the
-  // available() and read() functions.
-  uint8_t requestFrom(int address, int quantity, int stop) {
-    // TODO: implement
-    return 0; // number of bytes returned from the slave device
+  // Used by the master to request bytes from a slave device. The bytes may then
+  // be retrieved with the available() and read() functions.
+  uint8_t requestFrom(uint8_t address, size_t quantity, bool stop) {
+    assert(_didBegin);
+    assert(address > 0 && address < SLAVE_COUNT);
+    assert(quantity <= BUFFER_LENGTH);
+    in = &slaves[address];
+    // do we have enough data in the input buffer
+    if (quantity <= (in->misoBuffer).size()) { // enough data
+      in->misoSize = quantity;
+      return quantity;
+    } else { // not enough data
+      in->misoSize = 0;
+      in = nullptr;
+      return 0;
+    }
+  }
+  uint8_t requestFrom(uint8_t address, size_t quantity) {
+    return requestFrom(address, quantity, true);
   }
   uint8_t requestFrom(int address, int quantity) {
-    int stop = true;
-    return requestFrom(address, quantity, stop);
-  }
-  uint8_t requestFrom(uint8_t address, uint8_t quantity) {
-    return requestFrom((int)address, (int)quantity);
+    return requestFrom((uint8_t)address, (size_t)quantity);
   }
```

Comment:
While this is a legitimate function "overload" in that `size_t` and `int` are different types, I think that the C++ compiler ought to be able to make implicit conversions for us. Furthermore, and more importantly, if we don't provide all possible alternatives to the integer types, then there will be an ambiguity error. I found this when someone calls this function and passes in a byte. I think it would be best to just have one function and let the C++ compiler do the type cases. Thanks!

Reply:
Which function should remain?

Reply:
Sorry for the ambiguity! The function with the smaller types calls the function with the larger types. Let's remove the one with the smaller types (`uint8_t`).